src/Nethermind/Nethermind.Synchronization/Peers/PeerInfo.cs

```diff
@@ -158,6 +202,66 @@ private static string BuildContextString(AllocationContexts contexts)
             return $"{((contexts & AllocationContexts.Headers) == AllocationContexts.Headers ? "H" : " ")}{((contexts & AllocationContexts.Bodies) == AllocationContexts.Bodies ? "B" : " ")}{((contexts & AllocationContexts.Receipts) == AllocationContexts.Receipts ? "R" : " ")}{((contexts & AllocationContexts.State) == AllocationContexts.State ? "N" : " ")}{((contexts & AllocationContexts.Snap) == AllocationContexts.Snap ? "S" : " ")}";
         }
 
-        public override string ToString() => $"[{BuildContextString(AllocatedContexts)} ][{BuildContextString(SleepingContexts)} ]{SyncPeer}";
+        private string BuildSlotContextString()
+        {
+            StringBuilder ctxStringBuilder = new StringBuilder();
+
+            (AllocationContexts, char)[] stringOrders =
+            [
+                (AllocationContexts.Headers, 'h'),
+                (AllocationContexts.Bodies, 'b'),
+                (AllocationContexts.Receipts, 'r'),
+                (AllocationContexts.State, 'n'),
+                (AllocationContexts.Snap, 's'),
+            ];
+
+            foreach ((AllocationContexts ctx, char chRep) in stringOrders)
+            {
+                int count = AllocationSlots[ctx];
+                int allowances = _allocationAllowances[ctx];
+                if (count == allowances)
+                {
+                    ctxStringBuilder.Append(' ');
+                }
+                else if (count == 0)
+                {
+                    ctxStringBuilder.Append(Char.ToUpper(chRep));
+                }
+                else if (allowances - count == 1)
+                {
+                    ctxStringBuilder.Append(chRep);
+                }
+                else
+                {
+                    ctxStringBuilder.Append(allowances - count);
+                }
+            }
+
+            return ctxStringBuilder.ToString();
+        }
+
+        public override string ToString() => $"[{BuildSlotContextString()} ][{BuildContextString(SleepingContexts)} ]{SyncPeer}";
+
+        private AllocationContexts[] SeparateAllocationContexts(AllocationContexts contexts)
+        {
+            if (SeparatedContextsCache.TryGetValue(contexts, out AllocationContexts[] cachedContext))
+            {
+                return cachedContext;
+            }
+
+            cachedContext = Enum.GetValues<AllocationContexts>()
+                .Where(aCtx => IsOnlyOneContext(aCtx) && (contexts & aCtx) != 0)
+                .ToArray();
+
+            SeparatedContextsCache.TryAdd(contexts, cachedContext);
+            return cachedContext;
+        }
+
+        private static ConcurrentDictionary<AllocationContexts, AllocationContexts[]> SeparatedContextsCache = new ConcurrentDictionary<AllocationContexts, AllocationContexts[]>();
+
+        public static bool IsOnlyOneContext(AllocationContexts x)
+        {
+            return (x & (x - 1)) == 0;
+        }
```

Comment:
Do we need this and the cache, can't we just yield return from the method?

Reply:
But why? Its probably faster this way anyway. No alloc after cache is full.